src/SeoToolkit.Umbraco.Redirects.Core/Helpers/RedirectsImportHelper.cs

```diff
@@ -0,0 +1,248 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+using ExcelDataReader;
+using Microsoft.VisualBasic.FileIO;
+using SeoToolkit.Umbraco.Redirects.Core.Enumerators;
+using SeoToolkit.Umbraco.Redirects.Core.Interfaces;
+using SeoToolkit.Umbraco.Redirects.Core.Models.Business;
+using Umbraco.Cms.Core;
+using Umbraco.Cms.Core.Routing;
+using Umbraco.Cms.Core.Web;
+using Umbraco.Extensions;
+
+namespace SeoToolkit.Umbraco.Redirects.Core.Helpers;
+
+public class RedirectsImportHelper
+{
+    private Domain _selectedDomain;
+    private readonly IRedirectsService _redirectsService;
+    private readonly IUmbracoContextFactory _umbracoContextFactory;
+
+    public RedirectsImportHelper(IRedirectsService redirectsService, IUmbracoContextFactory umbracoContextFactory)
+    {
+        _redirectsService = redirectsService;
+        _umbracoContextFactory = umbracoContextFactory;
+    }
+
+    public Attempt<Dictionary<string,string>?, string> Validate(ImportRedirectsFileExtension fileExtension, MemoryStream memoryStream, string domain)
+    {
+        SetDomain(domain);
+        Attempt<Dictionary<string,string>, string> validationResult;
+        switch (fileExtension)
+        {
+            case ImportRedirectsFileExtension.Csv:
+                validationResult = ValidateCsv(memoryStream);
+                break;
+            case ImportRedirectsFileExtension.Excel:
+                validationResult = ValidateExcel(memoryStream);
+                break;
+            default:
+                return Attempt<Dictionary<string,string>?, string>.Fail("Invalid filetype, you may only use .csv or .xls", result: null);
+        }
+
+        if (validationResult.Success)
+        {
+            return Attempt<Dictionary<string,string>?, string>.Succeed(string.Empty, validationResult.Result);
+        }
+
+        return validationResult;
+    }
+
+    public Attempt<Dictionary<string,string>?, string> Import(ImportRedirectsFileExtension fileExtension, MemoryStream memoryStream, string domain)
+    {
+        SetDomain(domain);
+        var validation = Validate(fileExtension, memoryStream, domain);
+        if (validation is { Success: true, Result: not null } && validation.Result.Count != 0)
+        {
+            foreach (var entry in validation.Result)
+            {
+                SaveRedirect(entry);
+            }
+        }
+
+        return validation;
+    }
+
+    private bool UrlExists(string oldUrl)
```

Comment:
Just tested this locally and found an issue with this function. The search param on line 71 does a wildcard search on the From and To columns. So in my scenario I had a TO: /test1111 in the database and I wanted to add a FROM:/test11 but I wasn't allowed to add it because the system thought that the url already existed.

I think the search param is still fine, but we probably need to do a memory filter as well to only filter on exact matches on the from.

Reply:
Added!

Reply:
Good fine, will add!

Reply:
Have added it in RedirectsImportController.cs.
Haven't had time to test it thoroughly yet. Will do so in coming days, unless someone beats me to it ;)